### ISSUE TITLE
[v32.0-alpha] Satellite imagery: Download not working for offline

### ISSUE THREAD

**Issue by westnordost** [streetcomplete/StreetComplete#2772]
When downloading the map (either manual or automatic), the map tiles are also downloaded and cached in the upcoming major version.

However, this does currently not happen for the satellite imagery, when that map style is selected.

Furthermore, a link to the privacy statement for the ESRI tiles is missing.

### What to do?

`VectorTileProvider` should be refactored to look like this:

```kotlin
data class MapScene(
    val sceneFilePath: String,
    val sources: List<TileSource>
)

abstract class TileSource(
    val title: String,
    val maxZoom: Int,
    val copyrightText: String,
    val copyrightLink: String,
    val privacyStatementLink: String,
    val apiKey: String
) {
    abstract fun getTileUrl(zoom: Int, x: Int, y: Int): String
}
```

The `jawg` `MapScene` then has two sources, the jawg tiles and the esri tiles. Classes where the `TileSource` information is used need to be adapted. For example, if there are several tile sources which both require a `copyrightText` (that field and others could be made nullable if each none is required), they need to both turn up in the appropriate view.

The `MapTilesDownloader` will then accept a list of `TileSource`s (or a `MapScene` - not the best name perhaps) and download all the tiles for all the sources in the list.

This approach is not very economic with data as it always downloads the satellite data also if the style is not activated. So additional work is needed to only download the satellite imagery conditionally (=if the setting is active?). I guesst he most straightforward impelementation would be to exchange the `MapScene` when the setting is changed.

**Comment by smichel17** [comment]
It would also be nice if the switch were a little more prominent in the UI rather than buried in the settings. From my brief experience with the satellite data, I usually just wanted to tap-toggle it to check something (e.g. building outline) and then go back to map view, which was a pain with the switch buried in the settings.

(Happy to move this comment to a new issue if you don't want this one to be a grab bag of satellite imagery tweaks. Or not, I don't personally care about the satellite images very much.)

**Comment by peternewman** [comment]
I can imagine doing similar to this:
> I usually just wanted to tap-toggle it to check something (e.g. building outline) and then go back to map view

Which makes only having one available offline problematic:
> This approach is not very economic with data as it always downloads the satellite data also if the style is not activated. So additional work is needed to only download the satellite imagery conditionally (=if the setting is active?).

What about a three way download setting/second switch, so you can cache the active style or both?

**Comment by westnordost** [comment]
Yes, more settings, we need moorrre

Am 22. April 2021 12:14:56 MESZ schrieb Peter Newman ***@***.***>:
>I can imagine doing similar to this:
>> I usually just wanted to tap-toggle it to check something (e.g.
>building outline) and then go back to map view
>
>Which makes only having one available offline problematic:
>> This approach is not very economic with data as it always downloads
>the satellite data also if the style is not activated. So additional
>work is needed to only download the satellite imagery conditionally
>(=if the setting is active?).
>
>What about a three way download setting/second switch, so you can cache
>the active style or both?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/streetcomplete/StreetComplete/issues/2772#issuecomment-824716627

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by peternewman** [comment]
> Yes, more settings, we need moorrre

Or as an algorithmic, settings free model:

1. Download local imagery in chosen style
2. If user is not on WiFi/is on metered connection, go to END
3. Download opposite imagery, prioritising medium-high zooms in the chosen area
4. Continue to fetch progressively lower zooms/further out imagerry

**Comment by westnordost** [comment]
Anyway, a quick switch button like you see it on Google Maps is currently not planned. It would also be nice to hold off with such feature requests for some time, at least until v32.0 is out. That version is a huge endeavour and cramming even more and more features into this is not going to help (my stress level). As you know, the Prototype fund ended in February, yet I still find myself practically working on this app full-time. This is not sustainable.

I am going to hide this conversation, as this is all off topic.

**Comment by smichel17** [comment]
I feel I've not contributed as much code to SC as I would like (relative to amount of bikeshedding in PRs :sweat_smile:), and this looks like it should not be too much work, so I'm up to take a pass at it, unless @matkoniecz has already started.

**Comment by matkoniecz** [comment]
> so I'm up to take a pass at it, unless @matkoniecz has already started.

Sadly not yet started and likely will not today.

(though I am planning to work on it. Omplementing things and then dumping on @westnordost task of actually completing it is not a good strategy.

**Comment by matkoniecz** [comment]
So if @smichel17 wants: feel free to work on it! It would be appreciated.

**Comment by smichel17** [comment]
Ok. I will change the interface as suggested at the top, fix the compiler errors, and then open a draft PR and you can decide if you want to do anything else from there or if that's all that is needed.

**Comment by westnordost** [comment]
Wait wait, there are a few considerations that aren't in my original deliberations:
1. satellite imagery changes very infrequently. So, the satellite imagery needs a different cacheControl than the vector tiles, one where tiles are considered fresh for - I don't know - 6 months? Didn't think it through yet, but perhaps the information how long tiles are considered fresh and after what time they should be disregarded could be part of the `TileSource` instead of hardcoded into the `MapTilesDownloadCacheConfig`
2. if the satellite imagery is only fetched once and then not again for a very long time, I think probably no extra effort needs to go into conditionally not downloading satellite imagery but instead always download it. Depends on the size of the tiles too.

**Comment by smichel17** [comment]
1. Seems like this is a more-or-less independent chunk of the refactor / isn't a reason not to go ahead? It wasn't in `VectorTileProvider` before, so we can do this change after https://github.com/streetcomplete/StreetComplete/issues/2772#issuecomment-825555836?

   I'd also not mind waiting on this part because I am admittedly a bit fuzzy on how the cache actually works — fresh/stale vs maxAge, and how those values (and the max size) actually affect when cached tiles are used and when new ones are fetched. I can spend some time figuring it out, but I'd rather isolate "touching code I'm not as comfortable with" to a different commit.

2. Is this in reply to https://github.com/streetcomplete/StreetComplete/issues/2772#issuecomment-824806868?

**Comment by westnordost** [comment]
> It wasn't in VectorTileProvider before

Because it was the only tile provider.

> Is this in reply to #2772 (comment)?

No, it is a reply to 

> This approach is not very economic with data as it always downloads the satellite data also if the style is not activated. So additional work is needed to only download the satellite imagery conditionally (=if the setting is active?). I guesst he most straightforward impelementation would be to exchange the MapScene when the setting is changed.

**Comment by smichel17** [comment]
What's the relationship between Jawg and Esri? I thought they were independent, but the satellite scene/styles are inside the jawg folder and your suggested interface has the file path at the `MapScene` level (but maybe this just reflects the folder structure, not any other meaning).

Aside: I'm comfortable with manual dependency injection, but have not used Dagger or other DI frameworks before. I've read through enough of the Dagger docs now to understand how it works for our case, but not enough to understand the full available functionality.

**Comment by matkoniecz** [comment]
> What's the relationship between Jawg and Esri? 

Basically none.

> I thought they were independent, but the satellite scene/styles are inside the jawg folder and your suggested interface has the file path at the MapScene level (but maybe this just reflects the folder structure, not any other meaning).

Satellite style has also labels/roads displayed from Jawg vector tiles.

> satellite scene/styles are inside the jawg folder

jawg folder has jawg compatible map styles, Esri aerial is added as raster layer

And even aerial is also using vector data anyway

**Comment by westnordost** [comment]
> edit: It's not totally clear to me how @Inject constructor(private val x: Y) is saving much over constructor(private val x: Y = Y()) 🤷

Bit off topic, but: In that example, nothing really. In any (larger) application, you rather have a dependency **hierarchy**. A dependency injection framework lets you just state what you need and builds that dependency tree for you. Taking your example, just imagine `Y` also has dependencies. So for example, to construct a `Y`, you need to call `Y(A(),B(C(E()),D()))`. You definitely don't want to write this code in every place where you need a `Y`. Also, maybe you want that all classes that use a `Y` actually use the same instance of a `Y`.

**Comment by smichel17** [comment]
Yeah, it's tangential. DI came up at all because I was figuring out whether the `MapScene` or `TileSource` should be injected, and one of my weaknesses as a programmer is that I have a hard time black-boxing; it's important for me to understand the full stack pretty well (which is also a strength, depending on your viewpoint :upside_down_face:).

I still don't see why you couldn't do the same thing with default constructor params at each level:
`class Y constructor(public a: A = A(), public b: B = B())`,
`class B constructor(public c: C = C(), public d: D = D())`,
etc

**Comment by smichel17** [comment]
> Satellite style has also labels/roads displayed from Jawg vector tiles.

I guess this means that the api key is also the same between the tile sources? Should that be pulled up into `MapScene`?

Brainstorming on naming: maybe the top-level data class should still be named <code><del>Vector</del>TileProvider</code>, and the lower level is `RasterLayer`? This doesn't feel quite right yet, but closer. Should the base layer, which provides labels/roads, get special treatment? Maybe we don't need this abstraction yet, and we should just keep `VectorTileProvider` and add a `satelliteLayer: RasterTileProvider` property.

I'm just about done understanding and am going to start writing code soon, so maybe let's wait to continue this discussion until I open a draft PR. I'm leaning toward "just add a satelliteLayer property" for the initial draft and then we can discuss from there.

One other thing though: I want to understand the path that tiles take to get to Tangram. For the ESRI tiles, the download url is specified in the yaml scene file, which seems to be passed to Tangram directly. Josm tiles are instead…

- Somewhere along the way, we determine that we need to download more tiles and call `Downloader::download()` on that area (which calls `MapTilesDownloader::download()`)
 - The `MapTilesDownloader` gets the url from the `VectorTileProvider` and stores the results into the global `cacheConfig`
 - The tangram `mapView` is initialized with a http handler that uses this `cacheConfig`.

Does tangram will ever fetch tiles itself over http? I'm guessing that it does, and the above process is only used for offline tiles? In this case, where does Tangram get the jawg url from? `MapTilesDownloader::downloadTile` seems to be the only place where `VectorTileProvider::getTileUrl()` is called.

**Comment by matkoniecz** [comment]
> I guess this means that the api key is also the same between the tile sources?

Esri requires no API key.

**Comment by westnordost** [comment]
> I still don't see why you couldn't do the same thing with default constructor params at each level:
> class Y constructor(public a: A = A(), public b: B = B()),
> class B constructor(public c: C = C(), public d: D = D()),
> etc

- A could be an interface. `class AImpl : A` could be a class with certain library or runtime dependencies. You don't want these in `Y`. For example, `NoteDao` needs a `Database`. Currently the only implementation of that interface is `AndroidDatabase`, which has a runtime dependency on the Android environment. we don't want that `NoteDao` can only be used on Android.
- I mentioned already: Sometimes you want that there is only one instance of a certain dependency. For example, the `CountryBoundary` class needs 0.5s to 2s to load the data from disk. You don't want that every class that needs a `CountryBoundary` for one thing or another instantiates its own instance of that class and thus load the data from disk another time.

**Comment by smichel17** [comment]
> Satellite style has also labels/roads displayed from Jawg vector tiles.

So when using the ESRI layer, both the ESRI and Jawg `maxZoom`s are limiting factors, correct?

**Comment by westnordost** [comment]
On naming:

Within on style for tangram-es, you could have any number of tile sources. It is possible to combine different vector tile and raster tile sources in one style. Then, all these sources are used together. The map scene is like a CSS stylesheet: In the CSS stylesheet, you can refer to any number of image URLs (mediocre comparison).

**Comment by westnordost** [comment]
Jawg tiles are only available up til zoom 16 I think. This does not mean that you can't zoom in more, just the z16 tiles are the "highest resolution" tiles of Jawg.
ESRI raster tiles may be available till other max zoom levels, I didn't check.

**Comment by smichel17** [comment]
> Sometimes you want that there is only one instance of a certain dependency

```kt
import ...CountryBoundarySingleton
class A constructor(
    public countryBoundary: CountryBoundary = CountryBoundarySingleton,
)
```
I figure the code to initialize the (optionally lazy) singleton is roughly equivalent in length to the Dagger setup, with much less magic.

> runtime dependency on the Android environment

Okay, that makes sense. You could import the default implementation from a file that's dependent on your platform, but now that's more boilerplate that Dagger actually is saving you from.

**Comment by westnordost** [comment]
> One other thing though: I want to understand the path that tiles take to get to Tangram. For the ESRI tiles, the download url is specified in the yaml scene file, which seems to be passed to Tangram directly. Josm tiles are instead…

**All** map tile sources are actually specified in the scene yaml file(s). You also find the jawg one there, in `global.yaml`. Why is it then mentioned another time in the `VectorTileProvider` class? The reason is that this url is used by the `MapDataDownloader`.

tangram-es downloads the map tiles that are in view automatically (of course). But for StreetComplete offline-capability, we want to pre-download all the tiles in a given area when downloading a rectangle. So that is what the `MapDataDownloader` does: It downloads **the same** data as tangram-es would automatically. The "trick" here is that Streetcomplete tells tangram-es to use a certain cache directory (and config) and uses the exact same cache directory (and config) for the `MapDataDownloader`. So, things that are downloaded by this class can be used by tangram-es and the other way round.

**Comment by westnordost** [comment]
> import ...CountryBoundarySingleton
> class A constructor(
>     public countryBoundary: CountryBoundary = CountryBoundarySingleton,
> )
> 
> I figure the code to initialize the (optionally lazy) singleton is roughly equivalent in length to the Dagger setup, with much less magic.

But it is not the the class `A`s responsibility to decide about whether its countryBoundary should be a singleton or not.

**Comment by westnordost** [comment]
Also, the boilerplate you see in Dagger 2 in StreetComplete is partly due to Android. In "pure" JVM it would look cleaner.
A more convenient DI framework for Kotlin is https://insert-koin.io/

**Comment by smichel17** [comment]
Ah, I missed `global.yaml` :facepalm:

I found https://www.esri.com/en-us/privacy/privacy-statements/privacy-supplement for a privacy policy link, but I think this is for use of their website, not necessarily their api.

**Comment by smichel17** [comment]
Ah, good 'old Android Studio

![Memory: 14.4 (91%) of 15.6 GiB; Swap: 7.6 (64%) of 11.9 GiB](https://user-images.githubusercontent.com/6563664/115937042-614de580-a486-11eb-81c4-1f98da0fa2eb.png)

---

Anyway, aside, accessing the preferences seems a little boilerplate-y. Would you be interested in something like this? https://github.com/LibreShift/red-moon/blob/master/app/src/main/java/com/jmstudios/redmoon/model/Config.kt It's a `Config` singleton which uses the application context, so you can import it from anywhere and then access/set preference values like `Config.whateverValue`:

https://github.com/LibreShift/red-moon/blob/221e66728c6ca1838509ae55e7e7538706e5b87b/app/src/main/java/com/jmstudios/redmoon/schedule/ScheduleReceiver.kt#L59-L65

**Comment by FloEdelmann** [comment]
> How does one get [embedded code snippets](https://github.blog/2017-08-15-introducing-embedded-code-snippets/) to actually work? :/

See https://docs.github.com/en/github/managing-your-work-on-github/creating-a-permanent-link-to-a-code-snippet:

> This type of permanent link will render as a code snippet only in the repository it originated in. In other repositories, the permalink code snippet will render as a URL.

**Comment by westnordost** [comment]
@matkoniecz @smichel17 maybe we could back out the Satellite imagery feature from v32.0-alpha again and add it back as one package with #2794? I would feel a lot more confident with that, as currently v32.0-alpha is still quite unstable and some question marks left and I want to focus my efforts rather on that than attending to yet new (unstable) features.

**Comment by smichel17** [comment]
Sure. I'm working on it now so we can probably get it in for a release, but I think it's a large enough feature (from the user's perspective) to warrant its own release anyway.

**Comment by matkoniecz** [comment]
(1) Obviously, feel free to do this!

(2) I would prefer for it to be released, but I think that I can avoid any negative consequences of that, and chance that it would cause problems are tiny anyway

**Comment by westnordost** [comment]
Ok, I removed the option from the settings but left the rest of the code, hoping that it can be finished soon.

**Comment by westnordost** [comment]
I'll close this issue as this is now not an issue for v32-alpha anymore but will be implemented by that PR
